Source/MaestroVelPred.cpp

```diff
@@ -0,0 +1,1511 @@
+#include <Maestro.H>
+#include <MaestroHydro_F.H>
+
+using namespace amrex;
+
+#if (AMREX_SPACEDIM == 2)
+
+void 
+Maestro::VelPredInterface(const MFIter& mfi, 
+                          Array4<const Real> const utilde,
+                          Array4<const Real> const ufull,
+                          Array4<const Real> const utrans,
+                          Array4<const Real> const vtrans,
+                          Array4<const Real> const Imu,
+                          Array4<const Real> const Ipu,
+                          Array4<const Real> const Imv,
+                          Array4<const Real> const Ipv,
+                          Array4<Real> const ulx,
+                          Array4<Real> const urx,
+                          Array4<Real> const uimhx,
+                          Array4<Real> const uly,
+                          Array4<Real> const ury,
+                          Array4<Real> const uimhy,
+                          const Box& domainBox,
+                          const Real* dx)
+{
+    // timer for profiling
+    BL_PROFILE_VAR("Maestro::VelPredInterface()",VelPredInterface);
+
+    // NOTE: for ppm_type == 0, slopex == Ipu, slopey == Imv
+
+    ////////////////////////////////////
+    // Create u_{\i-\half\e_x}^x, etc.
+    ////////////////////////////////////
+
+    const Box& tileBox = mfi.tilebox();
+    const Box& obx = amrex::grow(tileBox, 1);
+    const Box& mxbx = amrex::growLo(obx, 0, -1);
+    const Box& mybx = amrex::growLo(obx, 1, -1);
+
+    Real rel_eps;
+    get_rel_eps(&rel_eps);
+
+    const Real dt2 = 0.5 * dt;
+
+    const Real hx = dx[0];
+    const Real hy = dx[1];
+
+    static const int ppm_type_local = ppm_type;
+
+    // x-direction
+    const int ilo = domainBox.loVect()[0];
+    const int ihi = domainBox.hiVect()[0];
+
+    int bclo = phys_bc[0];
+    int bchi = phys_bc[AMREX_SPACEDIM];
+
+    AMREX_PARALLEL_FOR_3D(mxbx, i, j, k, 
+    {
+        if (ppm_type_local == 0) {
+            Real maxu = max(0.0,ufull(i-1,j,k,0));
+            Real minu = min(0.0,ufull(i  ,j,k,0));
+            // extrapolate both components of velocity to left face
+            ulx(i,j,k,0) = utilde(i-1,j,k,0) + (0.5 - (dt2/hx)*maxu)*Ipu(i-1,j,k,0);
+            ulx(i,j,k,1) = utilde(i-1,j,k,1) + (0.5 - (dt2/hx)*maxu)*Ipu(i-1,j,k,1);
+            // extrapolate both components of velocity to right face
+            urx(i,j,k,0) = utilde(i  ,j,k,0) - (0.5 + (dt2/hx)*minu)*Ipu(i  ,j,k,0);
+            urx(i,j,k,1) = utilde(i  ,j,k,1) - (0.5 + (dt2/hx)*minu)*Ipu(i  ,j,k,1);
+        } else if (ppm_type_local == 1 || ppm_type_local == 2) {
+            // extrapolate both components of velocity to left face
+            ulx(i,j,k,0) = Ipu(i-1,j,k,0);
+            ulx(i,j,k,1) = Ipv(i-1,j,k,0);
```

Comment:
this should be `Ipv(i-1,j,k,1)`?  e.g. last component is wrong?

Reply:
I'm also confused as to which integral we should be using.  This is supposed to be putting things on x-edges, so that would suggest to me that we want the integral under the parabola in the x-direction over udt, so doesn't that mean that these should both actually be `Ipu`, and just the component different?
```
ulx(i,j,k,0) = Ipu(i-1,j,k,0);
ulx(u,j,k,1) = Ipu(i-1,j,k,1);
```

Reply:
okay, this is fine